File: Java Playground/02_String.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;

/**
 * String Problems - Complete Guide for Interview Preparation.
 *
 * <p>Core Concepts:
 * <ul>
 *   <li>Java String immutability — always use StringBuilder for concatenation in loops.</li>
 *   <li>s.charAt(i) not s[i]; String comparison: equals() not ==.</li>
 *   <li>int[26] for lowercase letter frequencies (faster than HashMap).</li>
 *   <li>Character.isLetterOrDigit(), Character.toLowerCase().</li>
 *   <li>Two-pointer technique for palindromes and validation.</li>
 * </ul>
 *
 * <p>Tricky Parts:
 * <ul>
 *   <li>String.substring(start, end) — end is exclusive.</li>
 *   <li>Anagrams: same character frequencies, not same order.</li>
 *   <li>Palindrome check: ignore non-alphanumeric and case-insensitive.</li>
 *   <li>KMP failure function prevents redundant character comparisons.</li>
 * </ul>
 *
 * @author Google Senior SWE Interview Playground
 */
public class String {

  private static final int ALPHABET_SIZE = 26;

  /**
   * Check if two strings are anagrams.
   *
   * <p>LeetCode 242 - Valid Anagram (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1) (fixed 26-char alphabet)
   *
   * @param s first string
   * @param t second string
   * @return true if t is an anagram of s
   * @throws IllegalArgumentException if either string is null
   */
  public static boolean isAnagram(java.lang.String s, java.lang.String t) {
    if (s == null || t == null) {
      throw new IllegalArgumentException("Input strings cannot be null");
    }

    if (s.length() != t.length()) {
      return false;
    }

    int[] charCount = new int[ALPHABET_SIZE];
    for (int i = 0; i < s.length(); i++) {
      charCount[s.charAt(i) - 'a']++;
      charCount[t.charAt(i) - 'a']--;
    }

    for (int count : charCount) {
      if (count != 0) {
        return false;
      }
    }

    return true;
  }

  /**
   * Group anagrams together.
   *
   * <p>LeetCode 49 - Group Anagrams (Medium)
   *
   * <p>Time Complexity: O(n * k log k) where k is max string length
   *
   * <p>Space Complexity: O(n * k)
   *
   * @param strs array of strings
   * @return list of anagram groups
   * @throws IllegalArgumentException if strs is null
   */
  public static List<List<java.lang.String>> groupAnagrams(java.lang.String[] strs) {
    if (strs == null) {
      throw new IllegalArgumentException("Input array cannot be null");
    }

    Map<java.lang.String, List<java.lang.String>> anagramMap = new HashMap<>();

    for (java.lang.String word : strs) {
      char[] chars = word.toCharArray();
      Arrays.sort(chars);
      java.lang.String key = new java.lang.String(chars);

      anagramMap.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
    }

    return new ArrayList<>(anagramMap.values());
  }

  /**
   * Find the length of longest substring without repeating characters.
   *
   * <p>LeetCode 3 - Longest Substring Without Repeating Characters (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(min(n, 26))
   *
   * @param s input string
   * @return length of longest substring without repeating chars
   * @throws IllegalArgumentException if s is null
   */
  public static int lengthOfLongestSubstring(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    Map<Character, Integer> lastSeen = new HashMap<>();
    int maxLength = 0;
    int left = 0;

    for (int right = 0; right < s.length(); right++) {
      char rightChar = s.charAt(right);

      if (lastSeen.containsKey(rightChar)) {
        left = Math.max(left, lastSeen.get(rightChar) + 1);
      }

      lastSeen.put(rightChar, right);
      maxLength = Math.max(maxLength, right - left + 1);
    }

    return maxLength;
  }

  /**
   * Minimum window substring containing all target characters.
   *
   * <p>LeetCode 76 - Minimum Window Substring (Hard)
   *
   * <p>Time Complexity: O(s + t), Space Complexity: O(1) (fixed ASCII)
   *
   * @param s haystack string
   * @param t needle string
   * @return minimum window substring or empty string if not found
   * @throws IllegalArgumentException if either string is null
   */
  public static java.lang.String minWindowSubstring(java.lang.String s,
      java.lang.String t) {
    if (s == null || t == null) {
      throw new IllegalArgumentException("Input strings cannot be null");
    }

    if (t.length() > s.length()) {
      return "";
    }

    Map<Character, Integer> targetCount = new HashMap<>();
    for (char c : t.toCharArray()) {
      targetCount.put(c, targetCount.getOrDefault(c, 0) + 1);
    }

    Map<Character, Integer> windowCount = new HashMap<>();
    int required = targetCount.size();
    int formed = 0;
    int left = 0;
    int minLength = Integer.MAX_VALUE;
    int minStart = 0;

    for (int right = 0; right < s.length(); right++) {
      char rightChar = s.charAt(right);
      windowCount.put(rightChar, windowCount.getOrDefault(rightChar, 0) + 1);

      if (targetCount.containsKey(rightChar) &&
          windowCount.get(rightChar).equals(targetCount.get(rightChar))) {
        formed++;
      }

      while (formed == required && left <= right) {
        if (right - left + 1 < minLength) {
          minLength = right - left + 1;
          minStart = left;
        }

        char leftChar = s.charAt(left);
        windowCount.put(leftChar, windowCount.get(leftChar) - 1);
        if (targetCount.containsKey(leftChar) &&
            windowCount.get(leftChar) < targetCount.get(leftChar)) {
          formed--;
        }

        left++;
      }
    }

    return minLength == Integer.MAX_VALUE ? "" : s.substring(minStart,
        minStart + minLength);
  }

  /**
   * Check if string is a valid palindrome (ignoring non-alphanumeric).
   *
   * <p>LeetCode 125 - Valid Palindrome (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1)
   *
   * @param s input string
   * @return true if s is a valid palindrome
   * @throws IllegalArgumentException if s is null
   */
  public static boolean isPalindrome(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    int left = 0;
    int right = s.length() - 1;

    while (left < right) {
      while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
        left++;
      }

      while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
        right--;
      }

      char leftChar = Character.toLowerCase(s.charAt(left));
      char rightChar = Character.toLowerCase(s.charAt(right));

      if (leftChar != rightChar) {
        return false;
      }

      left++;
      right--;
    }

    return true;
  }

  /**
   * Find longest palindromic substring using expand-around-center.
   *
   * <p>LeetCode 5 - Longest Palindromic Substring (Medium)
   *
   * <p>Time Complexity: O(n^2), Space Complexity: O(1)
   *
   * <p>Pattern: For each center, expand outward. Centers can be char (odd) or between
   * chars (even).
   *
   * @param s input string
   * @return longest palindromic substring
   * @throws IllegalArgumentException if s is null
   */
  public static java.lang.String longestPalindromicSubstring(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    if (s.length() < 2) {
      return s;
    }

    int maxStart = 0;
    int maxLength = 1;

    for (int i = 0; i < s.length(); i++) {
      int len1 = expandAroundCenter(s, i, i);
      int len2 = expandAroundCenter(s, i, i + 1);

      int len = Math.max(len1, len2);
      if (len > maxLength) {
        maxLength = len;
        maxStart = i - (len - 1) / 2;
      }
    }

    return s.substring(maxStart, maxStart + maxLength);
  }

  /**
   * Helper to expand around center for palindrome detection.
   *
   * @param s input string
   * @param left left center pointer
   * @param right right center pointer
   * @return length of palindrome centered at left/right
   */
  private static int expandAroundCenter(java.lang.String s, int left, int right) {
    while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
      left--;
      right++;
    }

    return right - left - 1;
  }

  /**
   * Check if string contains only valid parentheses.
   *
   * <p>LeetCode 20 - Valid Parentheses (Easy)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(n)
   *
   * @param s string containing parentheses
   * @return true if all parentheses are valid
   * @throws IllegalArgumentException if s is null
   */
  public static boolean isValid(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    Stack<Character> stack = new Stack<>();
    Map<Character, Character> pairs = new HashMap<>();
    pairs.put(')', '(');
    pairs.put('}', '{');
    pairs.put(']', '[');

    for (char c : s.toCharArray()) {
      if (pairs.containsValue(c)) {
        stack.push(c);
      } else if (pairs.containsKey(c)) {
        if (stack.isEmpty() || stack.pop() != pairs.get(c)) {
          return false;
        }
      }
    }

    return stack.isEmpty();
  }

  /**
   * Reverse words in string.
   *
   * <p>LeetCode 151 - Reverse Words in a String (Medium)
   *
   * <p>Time Complexity: O(n), Space Complexity: O(n)
   *
   * @param s input string with words separated by spaces
   * @return string with words in reverse order
   * @throws IllegalArgumentException if s is null
   */
  public static java.lang.String reverseWords(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    java.lang.String[] words = s.trim().split("\\s+");
    Collections.reverse(Arrays.asList(words));

    return java.lang.String.join(" ", words);
  }

  /**
   * Encode string using run-length encoding.
   *
   * <p>Time Complexity: O(n), Space Complexity: O(1) or O(n) for result
   *
   * @param s input string
   * @return encoded string
   * @throws IllegalArgumentException if s is null
   */
  public static java.lang.String encodeString(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    if (s.length() == 0) {
      return "";
    }

    StringBuilder encoded = new StringBuilder();
    int count = 1;

    for (int i = 1; i <= s.length(); i++) {
      if (i < s.length() && s.charAt(i) == s.charAt(i - 1)) {
        count++;
      } else {
        encoded.append(count).append(s.charAt(i - 1));
        count = 1;
      }
    }

    return encoded.toString();
  }

  /**
   * Decode run-length encoded string.
   *
   * <p>Time Complexity: O(n), Space Complexity: O(n)
   *
   * @param s encoded string
   * @return decoded string
   * @throws IllegalArgumentException if s is null
   */
  public static java.lang.String decodeString(java.lang.String s) {
    if (s == null) {
      throw new IllegalArgumentException("Input string cannot be null");
    }

    StringBuilder decoded = new StringBuilder();
    int i = 0;

    while (i < s.length()) {
      StringBuilder count = new StringBuilder();

      while (i < s.length() && Character.isDigit(s.charAt(i))) {
        count.append(s.charAt(i));
        i++;
      }

      int freq = count.length() > 0 ? Integer.parseInt(count.toString()) : 1;
      char c = s.charAt(i);

      for (int j = 0; j < freq; j++) {
        decoded.append(c);
      }

      i++;
    }

    return decoded.toString();
  }

  /**
   * KMP string matching: find first occurrence of pattern in text.
   *
   * <p>LeetCode 28 - Implement strStr() (Easy)
   *
   * <p>Time Complexity: O(m + n) where m is needle, n is haystack
   *
   * <p>Space Complexity: O(m) for failure function
   *
   * @param haystack text to search in
   * @param needle pattern to find
   * @return index of first occurrence or -1 if not found
   * @throws IllegalArgumentException if either argument is null
   */
  public static int strStr(java.lang.String haystack, java.lang.String needle) {
    if (haystack == null || needle == null) {
      throw new IllegalArgumentException("Input strings cannot be null");
    }

    if (needle.length() == 0) {
      return 0;
    }

    if (needle.length() > haystack.length()) {
      return -1;
    }

    int[] failureFunction = buildFailureFunction(needle);
    int needleIdx = 0;

    for (int haystackIdx = 0; haystackIdx < haystack.length(); haystackIdx++) {
      while (needleIdx > 0 && haystack.charAt(haystackIdx) !=
          needle.charAt(needleIdx)) {
        needleIdx = failureFunction[needleIdx - 1];
      }

      if (haystack.charAt(haystackIdx) == needle.charAt(needleIdx)) {
        needleIdx++;
      }

      if (needleIdx == needle.length()) {
        return haystackIdx - needle.length() + 1;
      }
    }

    return -1;
  }

  /**
   * Build KMP failure function (also called prefix function).
   *
   * @param pattern input pattern
   * @return failure function array
   */
  private static int[] buildFailureFunction(java.lang.String pattern) {
    int[] failure = new int[pattern.length()];
    int j = 0;

    for (int i = 1; i < pattern.length(); i++) {
      while (j > 0 && pattern.charAt(i) != pattern.charAt(j)) {
        j = failure[j - 1];
      }

      if (pattern.charAt(i) == pattern.charAt(j)) {
        j++;
      }

      failure[i] = j;
    }

    return failure;
  }

  /**
   * Main method with comprehensive test cases.
   *
   * @param args command-line arguments (unused)
   */
  public static void main(java.lang.String[] args) {
    System.out.println("=== String Problems ===\n");

    // Test isAnagram
    System.out.println("1. Valid Anagram:");
    boolean anagram = isAnagram("anagram", "nagaram");
    System.out.println("  Input: s=\"anagram\", t=\"nagaram\"");
    System.out.println("  Output: " + anagram + "\n");

    // Test groupAnagrams
    System.out.println("2. Group Anagrams:");
    java.lang.String[] strs = {"eat", "tea", "ate", "bat", "tab"};
    List<List<java.lang.String>> groups = groupAnagrams(strs);
    System.out.println("  Input: [\"eat\",\"tea\",\"ate\",\"bat\",\"tab\"]");
    System.out.println("  Output: " + groups + "\n");

    // Test lengthOfLongestSubstring
    System.out.println("3. Longest Substring Without Repeating Characters:");
    int longLen = lengthOfLongestSubstring("abcabcbb");
    System.out.println("  Input: \"abcabcbb\"");
    System.out.println("  Output: " + longLen + "\n");

    // Test isPalindrome
    System.out.println("4. Valid Palindrome:");
    boolean palindrome = isPalindrome("A man, a plan, a canal: Panama");
    System.out.println("  Input: \"A man, a plan, a canal: Panama\"");
    System.out.println("  Output: " + palindrome + "\n");

    // Test longestPalindromicSubstring
    System.out.println("5. Longest Palindromic Substring:");
    java.lang.String longestPal = longestPalindromicSubstring("babad");
    System.out.println("  Input: \"babad\"");
    System.out.println("  Output: \"" + longestPal + "\"\n");

    // Test isValid
    System.out.println("6. Valid Parentheses:");
    boolean valid = isValid("()[]{}");
    System.out.println("  Input: \"()[]{}\"");
    System.out.println("  Output: " + valid + "\n");

    // Test reverseWords
    System.out.println("7. Reverse Words in String:");
    java.lang.String reversed = reverseWords("  the   quick   brown  fox  ");
    System.out.println("  Input: \"  the   quick   brown  fox  \"");
    System.out.println("  Output: \"" + reversed + "\"\n");

    // Test encode/decode
    System.out.println("8. Encode/Decode String:");
    java.lang.String encoded = encodeString("aaabbbccc");
    System.out.println("  Encode \"aaabbbccc\": \"" + encoded + "\"");
    java.lang.String decoded = decodeString(encoded);
    System.out.println("  Decode back: \"" + decoded + "\"\n");

    // Test strStr
    System.out.println("9. String Matching (KMP):");
    int strIdx = strStr("hello", "ll");
    System.out.println("  Input: haystack=\"hello\", needle=\"ll\"");
    System.out.println("  Output: " + strIdx + "\n");
  }
}

File: C++ Playground/02_String.cpp
/// @file 02_String.cpp
/// @brief Strings Playground — frequency, sliding window, palindrome,
///        parentheses, word reversal, KMP string matching.
///
/// STRINGS GUIDE:
/// - std::string is mutable in C++; use += not + in loops
/// - std::string_view is non-owning read-only view; use for parameters
/// - Concatenation in loops: += is O(1) amortised; + is O(n)
/// - s[i] accesses char; std::string::npos for not found (size_t, not int)
/// - int[26] frequency array faster than unordered_map for lowercase letters
/// - Sliding window with int[128] for ASCII counts (avoid map overhead)
/// - Palindrome expand-around: check (i, i) and (i, i+1) for odd/even lengths
/// - KMP failure function: LPS[i] = longest proper prefix that is suffix

#include <algorithm>
#include <cctype>
#include <climits>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

namespace string_playground {

/// @brief Check if two strings are anagrams (LC 242 Easy).
/// @param s First string.
/// @param t Second string.
/// @return true if t is anagram of s (same characters, different order).
///
/// Example: s="anagram", t="nagaram" → true; s="ab", t="bb" → false
/// Constraints: 1 ≤ s.length, t.length ≤ 5×10⁴; lowercase English.
/// Time: O(n), Space: O(1) (fixed 26 chars)
/// Tricky: Count frequency of each char; if all match, anagrams.
[[nodiscard]] bool IsAnagram(std::string_view s,
                              std::string_view t) noexcept {
  if (s.length() != t.length()) return false;
  int char_count[26] = {0};
  for (char c : s) ++char_count[c - 'a'];
  for (char c : t) {
    --char_count[c - 'a'];
    if (char_count[c - 'a'] < 0) return false;
  }
  return true;
}

/// @brief Group anagrams together (LC 49 Medium).
/// @param strs Vector of strings.
/// @return Vector of vectors, each subvector contains anagrams of each other.
///
/// Example: strs=["eat","tea","tan","ate","nat","bat"] →
///          [["eat","tea","ate"],["tan","nat"],["bat"]]
/// Constraints: 1 ≤ n ≤ 10⁴; 1 ≤ strs[i].length ≤ 100.
/// Time: O(n * k log k) where k = avg string length (sorting chars).
/// Space: O(n * k) for map
/// Tricky: Sort chars in each string to get canonical form as key.
[[nodiscard]] std::vector<std::vector<std::string>> GroupAnagrams(
    const std::vector<std::string>& strs) {
  std::unordered_map<std::string, std::vector<std::string>> anagram_groups;
  for (const auto& str : strs) {
    std::string sorted_str = str;
    std::sort(sorted_str.begin(), sorted_str.end());
    anagram_groups[sorted_str].emplace_back(str);
  }

  std::vector<std::vector<std::string>> result;
  for (auto& group : anagram_groups) {
    result.emplace_back(std::move(group.second));
  }
  return result;
}

/// @brief Length of longest substring without repeating characters (LC 3 Med).
/// @param s Input string.
/// @return Length of longest substring with all unique characters.
///
/// Example: s="abcabcbb" → 3 (substring "abc"); s="au" → 2; s="dvdf" → 3
/// Constraints: 0 ≤ s.length ≤ 5×10⁴; lowercase, uppercase, digits, symbols.
/// Time: O(n), Space: O(min(n, charset_size))
/// Tricky: Sliding window with char→last_seen_index map; when char repeats,
///         move left past previous occurrence.
[[nodiscard]] int LengthOfLongestSubstring(std::string_view s) noexcept {
  std::unordered_map<char, int> char_index;
  int left = 0, max_len = 0;

  for (int right = 0; right < static_cast<int>(s.length()); ++right) {
    if (char_index.count(s[right]) && char_index[s[right]] >= left) {
      left = char_index[s[right]] + 1;
    }
    char_index[s[right]] = right;
    max_len = std::max(max_len, right - left + 1);
  }
  return max_len;
}

/// @brief Minimum window substring: find smallest substring containing
///        all chars from t (LC 76 Hard).
/// @param s Haystack string.
/// @param t Pattern string.
/// @return Smallest substring of s containing all chars of t (with counts).
///
/// Example: s="ADOBECODEBANC", t="ABC" → "BANC"
/// Constraints: 1 ≤ s.length, t.length ≤ 10⁵; any chars.
/// Time: O(n + m) where n = s.len, m = t.len
/// Space: O(1) (fixed 128 ASCII chars)
/// Tricky: Track required char counts; expand right until window valid;
///         shrink left while valid to minimize size; track best window.
[[nodiscard]] std::string MinWindowSubstring(std::string_view s,
                                              std::string_view t) {
  if (s.length() < t.length()) return "";

  int required[128] = {0}, window[128] = {0};
  for (char c : t) ++required[static_cast<unsigned char>(c)];

  int left = 0, formed = 0, required_count = 0;
  for (int i = 0; i < 128; ++i) {
    if (required[i] > 0) ++required_count;
  }

  int result_len = INT_MAX, result_left = 0;

  for (int right = 0; right < static_cast<int>(s.length()); ++right) {
    unsigned char c = s[right];
    ++window[c];
    if (required[c] > 0 && window[c] == required[c]) {
      --formed;
    }

    while (formed == 0 && left <= right) {
      if (right - left + 1 < result_len) {
        result_len = right - left + 1;
        result_left = left;
      }

      unsigned char left_char = s[left];
      --window[left_char];
      if (required[left_char] > 0 && window[left_char] < required[left_char]) {
        ++formed;
      }
      ++left;
    }
  }

  return result_len == INT_MAX ? ""
                               : std::string(s.substr(result_left, result_len));
}

/// @brief Check if string is valid palindrome (alphanumeric only) (LC 125 Easy).
/// @param s Input string (may contain spaces, punctuation).
/// @return true if alphanumeric chars form palindrome.
///
/// Example: s="A man, a plan, a canal: Panama" → true
/// Constraints: 1 ≤ s.length ≤ 2×10⁵; any ASCII chars.
/// Time: O(n), Space: O(1)
/// Tricky: Skip non-alphanumeric; compare lowercase versions.
[[nodiscard]] bool IsPalindrome(std::string_view s) noexcept {
  int left = 0, right = static_cast<int>(s.length()) - 1;
  while (left < right) {
    while (left < right && !std::isalnum(s[left])) ++left;
    while (left < right && !std::isalnum(s[right])) --right;

    if (std::tolower(s[left]) != std::tolower(s[right])) {
      return false;
    }
    ++left;
    --right;
  }
  return true;
}

/// @brief Longest palindromic substring: expand around center (LC 5 Medium).
/// @param s Input string.
/// @return Longest substring that is palindrome.
///
/// Example: s="babad" → "bab" or "aba"; s="cbbd" → "bb"
/// Constraints: 1 ≤ s.length ≤ 1000; lowercase English.
/// Time: O(n²), Space: O(1)
/// Tricky: For each center (odd-length: single char; even-length: between
///         chars), expand outwards while chars match. Track longest.
[[nodiscard]] std::string LongestPalindromicSubstring(
    std::string_view s) {
  if (s.empty()) return "";

  auto expand_around_center = [&](int left, int right)
      -> std::pair<int, int> {
    while (left >= 0 && right < static_cast<int>(s.length()) &&
           s[left] == s[right]) {
      --left;
      ++right;
    }
    return {left + 1, right - 1};
  };

  int max_start = 0, max_end = 0;
  for (int i = 0; i < static_cast<int>(s.length()); ++i) {
    auto [start1, end1] = expand_around_center(i, i);
    if (end1 - start1 > max_end - max_start) {
      max_start = start1;
      max_end = end1;
    }

    if (i + 1 < static_cast<int>(s.length())) {
      auto [start2, end2] = expand_around_center(i, i + 1);
      if (end2 - start2 > max_end - max_start) {
        max_start = start2;
        max_end = end2;
      }
    }
  }
  return std::string(s.substr(max_start, max_end - max_start + 1));
}

/// @brief Valid parentheses: check balanced brackets (LC 20 Easy).
/// @param s String of parens/brackets/braces.
/// @return true if valid (balanced and properly ordered).
///
/// Example: s="()" → true; s="(){}" → true; s="({[]})" → true; s="({)}" →
///          false
/// Constraints: 1 ≤ s.length ≤ 10⁴; only '(){}[]'.
/// Time: O(n), Space: O(n) (stack)
/// Tricky: Push open chars; pop and match on close; stack empty at end.
[[nodiscard]] bool IsValid(std::string_view s) {
  std::string stack;
  for (char c : s) {
    if (c == '(' || c == '{' || c == '[') {
      stack.push_back(c);
    } else {
      if (stack.empty()) return false;
      char top = stack.back();
      if ((c == ')' && top == '(') || (c == '}' && top == '{') ||
          (c == ']' && top == '[')) {
        stack.pop_back();
      } else {
        return false;
      }
    }
  }
  return stack.empty();
}

/// @brief Reverse words in string: "the sky is blue" → "blue is sky the"
///        (LC 151 Medium).
/// @param s Input string with words separated by spaces.
/// @return String with words in reverse order, single spaces.
///
/// Example: s="  Hello World  " → "World Hello"; s="  Bob    Loves  Alice   "
///          → "Alice Loves Bob"
/// Constraints: 1 ≤ s.length ≤ 10⁴; multiple spaces OK.
/// Time: O(n), Space: O(1) or O(n) depending on string mutability.
/// Tricky: Skip leading/trailing spaces; split by spaces; build result
///         in reverse order.
[[nodiscard]] std::string ReverseWords(std::string_view s) {
  std::vector<std::string> words;
  std::string word;

  for (char c : s) {
    if (c != ' ') {
      word += c;
    } else if (!word.empty()) {
      words.emplace_back(std::move(word));
      word.clear();
    }
  }
  if (!word.empty()) {
    words.emplace_back(std::move(word));
  }

  std::string result;
  for (int i = static_cast<int>(words.size()) - 1; i >= 0; --i) {
    result += words[i];
    if (i > 0) result += " ";
  }
  return result;
}

/// @brief KMP string matching: find pattern in text (LC 28 Medium).
/// @param haystack Text string.
/// @param needle Pattern string.
/// @return Index of first occurrence of needle in haystack, or -1.
///
/// Example: haystack="sadbutsad", needle="sad" → 0
/// Constraints: 1 ≤ needle.length ≤ haystack.length ≤ 10⁴.
/// Time: O(n + m) where n = haystack.len, m = needle.len
/// Space: O(m) for LPS array
/// Tricky: Build LPS (Longest Proper Prefix that is Suffix) array for
///         pattern; use it to skip redundant comparisons.
[[nodiscard]] int StrStr(std::string_view haystack,
                          std::string_view needle) noexcept {
  if (needle.empty()) return 0;
  if (haystack.length() < needle.length()) return -1;

  std::vector<int> lps(needle.length(), 0);
  int j = 0;
  for (int i = 1; i < static_cast<int>(needle.length()); ++i) {
    while (j > 0 && needle[i] != needle[j]) {
      j = lps[j - 1];
    }
    if (needle[i] == needle[j]) {
      ++j;
    }
    lps[i] = j;
  }

  j = 0;
  for (int i = 0; i < static_cast<int>(haystack.length()); ++i) {
    while (j > 0 && haystack[i] != needle[j]) {
      j = lps[j - 1];
    }
    if (haystack[i] == needle[j]) {
      ++j;
    }
    if (j == static_cast<int>(needle.length())) {
      return i - static_cast<int>(needle.length()) + 1;
    }
  }
  return -1;
}

}  // namespace string_playground

int main() {
  std::cout << "=== STRING PLAYGROUND ===\n\n";

  bool is_anagram =
      string_playground::IsAnagram("anagram", "nagaram");
  std::cout << "IsAnagram(\"anagram\", \"nagaram\"): "
            << (is_anagram ? "true" : "false") << "\n";

  std::vector<std::string> words = {"eat", "tea", "tan", "ate", "nat", "bat"};
  auto grouped = string_playground::GroupAnagrams(words);
  std::cout << "GroupAnagrams(...): " << grouped.size() << " groups\n";

  int longest_substring =
      string_playground::LengthOfLongestSubstring("abcabcbb");
  std::cout << "LengthOfLongestSubstring(\"abcabcbb\"): " << longest_substring
            << "\n";

  std::string min_window =
      string_playground::MinWindowSubstring("ADOBECODEBANC", "ABC");
  std::cout << "MinWindowSubstring(\"ADOBECODEBANC\", \"ABC\"): \"" << min_window
            << "\"\n";

  bool is_palindrome =
      string_playground::IsPalindrome("A man, a plan, a canal: Panama");
  std::cout << "IsPalindrome(\"A man, a plan, a canal: Panama\"): "
            << (is_palindrome ? "true" : "false") << "\n";

  std::string longest_palin =
      string_playground::LongestPalindromicSubstring("babad");
  std::cout << "LongestPalindromicSubstring(\"babad\"): \"" << longest_palin
            << "\"\n";

  bool is_valid = string_playground::IsValid("()[]{}");
  std::cout << "IsValid(\"()[]{}\"); " << (is_valid ? "true" : "false")
            << "\n";

  std::string reversed =
      string_playground::ReverseWords("  Hello World  ");
  std::cout << "ReverseWords(\"  Hello World  \"): \"" << reversed << "\"\n";

  int str_index = string_playground::StrStr("sadbutsad", "sad");
  std::cout << "StrStr(\"sadbutsad\", \"sad\"): " << str_index << "\n";

  return 0;
}
